### ISSUE TITLE
Pkg.test() environment causes compile failures

### ISSUE THREAD

**Issue by MikeInnes** [JuliaGPU/CUDAnative.jl#98]
In CuArrays I see a couple of errors under `Pkg.test()` that don't come up under `julia runtests.jl`.

```julia
Indexing: Error During Test
  Test threw an exception of type ErrorException
  Expression: collect(xs[1:2, 2]) == (collect(xs))[1:2, 2]
  error compiling index_kernel: emit_allocobj for /home/mike/.julia/v0.6/CuArrays/src/indexing.jl:37 requires the dynamic_alloc language feature, which is disabled

PermuteDims: Error During Test
  Test threw an exception of type CUDAdrv.CuError
  Expression: collect(permutedims(xs, (2, 1))) == permutedims(collect(xs), (2, 1))
  ptxas fatal   : Unresolved extern function 'jl_bounds_error_unboxed_int'
  CUDA error: no kernel image is available for execution on the device (code #209, ERROR_NO_BINARY_FOR_GPU)
```

I assume that this must be due to either things like forced bounds checks or lower optimisation levels.

**Comment by maleadt** [comment]
Do you require `--check-bounds=no`? Because [Pkg.test](https://github.com/JuliaLang/julia/blob/fe8a4413b72b8b9cb9a826c7d716646ebc3cc7cb/base/pkg/entry.jl#L715-L723) doesn't seem to call `runtests` any differently. In that case, can't you implement a GPU-compatible `boundscheck`?

**Comment by MikeInnes** [comment]
Nope, no special requirements in my code -- and I haven't implemented any boundscheck logic since that's all handled by base.

**Comment by maleadt** [comment]
Culprit is `--check-bounds=yes`. MWE:
```julia
using CUDAnative
CUDAnative.code_llvm(getindex, (UnitRange{Int64},Int64))
```

Apparently `UnitRange` has a bounds check, which now result in invalid code (probably because of the outlining of `throw_boundserror`). Wasn't noticed on `CuDeviceArray` because bounds-checks were disabled there because of #4.

wrt. CuArrays, not sure why the generous `@inbound`s in `index_kernel` don't apply to this case already, maybe there's an `@propagate_inbounds` missing somewhere?

**Comment by MikeInnes** [comment]
Ok, so the problem isn't range indexing, it's that `--check-bounds=yes` ignores `@inbounds` completely. Your test case fails for `getindex(::Int, ::Int)` too. This probably should have been obvious, but I was thrown by the fact that we didn't get indexing errors for all arrays (but, obviously, CuDeviceArray indexing is device compatible).

It would be great to fix this on the CUDAnative side by unsetting this option when compiling, but for now I'll find a way to ignore that flag in tests.

**Comment by maleadt** [comment]
> CuDeviceArray indexing is device compatible

as in, it does a `return nothing` 😄 

So yeah, it needs a fix to the CUDAnative JIT where outlined `throw(BoundsError)` (which is a thing in 0.6 to prevent a gcframe in the parent) aren't compiled properly.

**Comment by maleadt** [comment]
CUDAnative works with bounds checks now.
https://github.com/JuliaGPU/CUDAnative.jl/pull/256
